### PR TITLE
Deprecate `LightningModule.use_amp`

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/12307#issuecomment-1065903429

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/starter/core_guide.rst [hunk 1]
```diff
@@ -944,15 +944,15 @@ default implementation
 
 .. testcode::
 
-    def backward(self, use_amp, loss, optimizer):
```
Comment (ananthsub) [inline docs/source/starter/core_guide.rst:947]: This shouldn't be deprecated. It's different from the attribute in the lightning module.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -164,6 +164,7 @@ def _copy_trainer_model_properties(self, model):
 
         for m in [model, ref_model]:
             m.trainer = proxy(self.trainer)
+            # Remove next line in v1.8
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/data_connector.py:167]: Be more specific about use_amp in case the code here changes between now and v1.8

File: docs/source/starter/core_guide.rst (hunk 1 above)
Comment (daniellepintz) [reply docs/source/starter/core_guide.rst:947]: Right, I know it's different but thought we didn't need this here as well. Will put back

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1966,6 +1965,23 @@ def model_size(self) -> float:
...
+        rank_zero_deprecation(
+            "`LightningModule.use_amp` was deprecated in v1.6 and will be removed in v1.8."
+            " Please use `Trainer.amp_backend`.",
+            stacklevel=5,
+        )
+        return self._use_amp
+
+    @use_amp.setter
+    def use_amp(self, use_amp: bool) -> None:
+        self._use_amp = use_amp
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1983]: add a deprecation warning here too.

the trainer can then set the attribute using `_use_amp` to avoid the deprecation warning being logged.

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1966,6 +1965,23 @@ def model_size(self) -> float:
             )
         return get_model_size_mb(self)
 
+    @property
+    def use_amp(self) -> bool:
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1969]: properties need to be added here so theyre not exported to torchscript: https://github.com/PyTorchLightning/pytorch-lightning/blob/7ee690758ccad7f702460d056f6369c1d4371a46/pytorch_lightning/core/lightning.py#L69-L87

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (daniellepintz) [reply pytorch_lightning/core/lightning.py:1983]: > the trainer can then set the attribute using `_use_amp` to avoid the deprecation warning being logged.

Actually we don't even need this setter then

Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1983]: theoretically users could set `use_amp` as its a regular attribute now, so the setter + deprecation warning is needed for completeness.

File: tests/deprecated_api/test_remove_1-8.py [hunk 5]
```diff
@@ -817,3 +817,10 @@ def on_load_checkpoint(self, checkpoint):
         " v1.6 and will be removed in v1.8. Use `load_state_dict` instead."
     ):
         _check_datamodule_checkpoint_hooks(trainer)
+
+
+def test_v1_8_0_lightning_module_use_amp():
+    model = BoringModel()
+    with pytest.deprecated_call(match="`LightningModule.use_amp` was deprecated in v1.6"):
+        _ = model.use_amp
+        model.use_amp = False
```
[the remark this answers is not in the document]
Comment (daniellepintz) [reply tests/deprecated_api/test_remove_1-8.py:826]: oops sorry, fixed!

**Review by akihironitta — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1970]: Yes @daniellepintz I ran into this too a while ago: #10092

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1970]: Correct. It's bad.

You could try pushing https://github.com/pytorch/pytorch/issues/67146 internally.

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by daniellepintz** [comment]
Ready to be merged
